File: src/lib/wb/barcode.ts
import crypto from "node:crypto";

/**
 * Generate a valid EAN-13 barcode (with correct check digit). WB accepts
 * seller-provided barcodes as a size's SKU. We use the "200" in-store prefix
 * range which is reserved for internal/private use.
 */
export function generateEan13(): string {
  // 12 digits: "20" + 10 random, then compute the 13th check digit
  let digits = "20";
  const rnd = crypto.randomBytes(10);
  for (let i = 0; i < 10; i++) digits += String(rnd[i] % 10);
  return digits + ean13CheckDigit(digits);
}

function ean13CheckDigit(twelve: string): string {
  let sum = 0;
  for (let i = 0; i < 12; i++) {
    const d = twelve.charCodeAt(i) - 48;
    sum += i % 2 === 0 ? d : d * 3;
  }
  const check = (10 - (sum % 10)) % 10;
  return String(check);
}

File: src-tauri/src/wb/barcode.rs
#![allow(dead_code)]
//! EAN-13 barcode generation (ported from src/lib/wb/barcode.ts).
//! WB accepts a seller-provided barcode as a size's SKU. "20" is the in-store
//! prefix range reserved for internal/private use.

use rand::RngCore;

pub fn generate_ean13() -> String {
    let mut digits = String::from("20");
    let mut rnd = [0u8; 10];
    rand::thread_rng().fill_bytes(&mut rnd);
    for b in rnd.iter() {
        digits.push((b'0' + (b % 10)) as char);
    }
    let check = ean13_check_digit(&digits);
    format!("{}{}", digits, check)
}

fn ean13_check_digit(twelve: &str) -> u32 {
    let mut sum = 0u32;
    for (i, c) in twelve.bytes().enumerate() {
        let d = (c - b'0') as u32;
        sum += if i % 2 == 0 { d } else { d * 3 };
    }
    (10 - (sum % 10)) % 10
}
